extension/src/popup/helpers/useValidateTransactionMemo.ts

```diff
@@ -0,0 +1,271 @@
+};
+
+/**
+ * Hook to validate transaction memos for addresses that require them
+ *
+ * This hook checks if a transaction destination address requires a memo by:
+ * 1. Checking a cached list of memo-required addresses from StellarExpert API
+ * 2. Falling back to Stellar SDK's checkMemoRequired method
+ * 3. Only validating on mainnet when memo validation is enabled in preferences
+ *
+ * @param {string | null | undefined} incomingXdr - The transaction XDR string to validate
+ * @returns {Object} Validation state and results
+ * @returns {boolean} returns.isMemoMissing - Whether a required memo is missing
+ * @returns {boolean} returns.isValidatingMemo - Whether validation is currently in progress
+ *
+ * @example
+ * ```tsx
+ * const { isMemoMissing, isValidatingMemo } = useValidateTransactionMemo(transactionXDR);
+ *
+ * if (isMemoMissing && !isValidatingMemo) {
+ *   // Show warning that memo is required
+ * }
+ * ```
+ */
+export const useValidateTransactionMemo = (incomingXdr?: string | null) => {
+  const networkDetails = useSelector(settingsNetworkDetailsSelector);
+  const { isMemoValidationEnabled } = useSelector(settingsPreferencesSelector);
+  const activePublicKey = useSelector(publicKeySelector);
+  const [isValidatingMemo, setIsValidatingMemo] = useState(false);
+  const [localTransaction, setLocalTransaction] = useState<ReturnType<
+    typeof TransactionBuilder.fromXDR
+  > | null>(null);
+  const [memoRequiredAccounts, setMemoRequiredAccounts] = useState<
+    MemoRequiredAccount[]
+  >([]);
+
+  const xdr = useMemo(() => incomingXdr, [incomingXdr]);
+
+  /**
+   * Determines if memo validation should be performed
+   * Only validates on mainnet when the feature is enabled in preferences
+   */
+  const shouldValidateMemo = useMemo(
+    () => !!(isMemoValidationEnabled && isMainnet(networkDetails)),
+    [isMemoValidationEnabled, networkDetails],
+  );
+
+  // Start with true to prevent button from being enabled before validation completes
+  // Reset to true whenever XDR changes to ensure we re-validate
+  const [isMemoMissing, setIsMemoMissing] = useState(true);
+
+  /**
+   * Effect to fetch memo-required accounts from cache
+   */
+  useEffect(() => {
+    if (!shouldValidateMemo) {
+      return;
+    }
+
+    const fetchMemoRequiredAccounts = async () => {
+      if (!activePublicKey) {
+        return;
+      }
+      try {
+        const response = await internalGetMemoRequiredAccounts({
+          activePublicKey,
+        });
+        if (response && !(response instanceof Error)) {
+          setMemoRequiredAccounts(response.memoRequiredAccounts || []);
+        }
+      } catch (error) {
+        console.error("Error fetching memo-required accounts:", error);
+      }
+    };
+
+    fetchMemoRequiredAccounts();
+  }, [shouldValidateMemo, activePublicKey]);
+
+  /**
+   * Effect to parse XDR and set initial memo validation state
+   * Runs when XDR, network, or validation settings change
+   */
+  useEffect(() => {
```

Comment:
I think you should be able to consolidate these next 2 `useEffect` calls - they seem duplicative. You should be able to just call `checkIsMemoRequired  after you build the xdr and set `hasMemo`